PurchasesCoreSwift/Public/PurchaserInfo.swift

```diff
@@ -0,0 +1,317 @@
+     */
+    @objc public let originalPurchaseDate: Date?
+
+    /**
+    Returns the build number (in iOS) or the marketing version (in macOS) for the version of the application when the user bought the app.
+    This corresponds to the value of CFBundleVersion (in iOS) or CFBundleShortVersionString (in macOS) in the Info.plist file when the purchase was originally made.
+    Use this for grandfathering users when migrating to subscriptions.
+
+     
+     @note This can be nil, see -[RCPurchases restoreTransactionsForAppStore:]
+     */
+    @objc public let originalApplicationVersion: String?
+
+    // TODO after migration make this internal
+    @objc public let schemaVersion: String?
+
+    private let expirationDatesByProduct: [String: Date?]
+
+    private let purchaseDatesByProduct: [String: Date?]
+
+    private let originalData: [AnyHashable: Any]
+
+    @objc public init?(data: [AnyHashable: Any]) {
+        guard let subscriberObject = data["subscriber"] as? [String: Any],
+              let subscriberData = SubscriberData.init(subscriberData: subscriberObject)
+            else {
+            return nil
+        }
+
+        self.originalData = data
+        self.schemaVersion = data["schema_version"] as? String
+
+        guard let requestDateString = data["request_date"] as? String,
+              let formattedRequestDate = DateFormatter.dateFormatter.date(from: requestDateString) else {
+            return nil
+        }
+        self.requestDate = formattedRequestDate
+
+        self.originalPurchaseDate = subscriberData.originalPurchaseDate
+        self.firstSeen = subscriberData.firstSeen
+        self.originalAppUserId = subscriberData.originalAppUserId
+        self.managementURL = subscriberData.managementURL
+        self.originalApplicationVersion = subscriberData.originalApplicationVersion
+
+        self.nonConsumablePurchases = Set(subscriberData.nonSubscriptions.keys)
+        self.nonSubscriptionTransactions = subscriberData.nonSubscriptionTransactions
+
+        self.entitlements = EntitlementInfos.init(entitlementsData: subscriberData.entitlements, purchasesData: subscriberData.allPurchases, dateFormatter: DateFormatter.dateFormatter, requestDate: requestDate)
+
+        self.expirationDatesByProduct = subscriberData.expirationDatesByProduct
+        self.purchaseDatesByProduct = subscriberData.purchaseDatesByProduct
+    }
+
+    // TODO after migration make this internal
+    @objc public class func currentSchemaVersion() -> String { "2" }
+
+    /**
+     Get the expiration date for a given product identifier. You should use Entitlements though!
+    
+     @param forProductIdentifier Product identifier for product
+    
+     @return The expiration date for `productIdentifier`, `nil` if product never purchased
+     */
+    @objc public func expirationDate(forProductIdentifier: String) -> Date? { expirationDatesByProduct[forProductIdentifier] ?? nil }
```

Comment:
that's what i thought should be the case (for all of these), but when i remove `?? nil` i get this error:
`Value of optional type 'Date??' must be unwrapped to a value of type 'Date?'`. I think since it's possible the dictionary just doesn't contain a value with that key?